### PR TITLE
[STUD-575][Enhancement]: Implement general service disruption banner across NEWM Studio site

### PR DESCRIPTION
**Pull request by ITurres**
# Pull Request — Issue [STUD-575](https://projectnewm.atlassian.net/browse/STUD-575)

## Overview

> This PR implements a general service disruption banner across the NEWM Studio site to inform users about the discontinuation of NEWM Studio distribution and Stream Token sales. The banner is conditionally displayed based on a LaunchDarkly feature flag (`webStudioDisableDistributionAndSales`) and appears fixed at the top of the page on desktop viewports.  
> **Context**: This change was necessary to communicate important service updates to users in a prominent, non-intrusive manner. The banner provides a clear message with a link to the official statement, ensuring users are informed about changes affecting their releases.

---

## Files Summary

> **Total Changes:** 20 files  
> _(Counts of added, modified, and deleted files can be seen in the PR diff view.)_

<details>
<summary>Added (4)</summary>

- **`apps/studio/src/components/InfoBanner.tsx`**
  - — New component that renders a service disruption banner using the shared `Banner` component. Conditionally displays based on `webStudioDisableDistributionAndSales` LaunchDarkly flag. Uses `OfficialStatementCTA` component for the banner message. Fixed positioning on desktop viewports.
- **`apps/studio/src/components/OfficialStatementCTA.tsx`**
  - — Reusable component that renders the official statement CTA message with a link to the official statement page. Encapsulates the message text and link styling to avoid duplication across multiple locations (e.g. pages declared in tickets STUD-576, STUD-577 and STUD-578). Accepts configurable `href` and `linkText` props while maintaining consistent styling.
- **`packages/components/src/lib/test/Banner.test.tsx`**
  - — Comprehensive unit test suite for the `Banner` component with 22 test cases covering rendering, props, styling, and edge cases. Tests follow Testing Library best practices and avoid direct DOM node access.
- **`packages/components/src/lib/Banner.tsx`**
  - — Enhanced component API by replacing `titleFontWeight` prop with more flexible `titleSx` prop (SxProps<Theme>) and added `sx` prop for root Box styling. Added `data-testid="banner"` for testing purposes. This provides better flexibility and aligns with MUI patterns used throughout the codebase.

</details>

<details>
<summary>Modified (16)</summary>

- **`apps/studio/src/components/Background.tsx`**
  - — Added conditional top offset to the main content overlay when `webStudioDisableDistributionAndSales` is enabled: `top: theme.spacing(2)` for viewports between 960px and xl so content clears the fixed banner. Offset is applied via an `applyOffset` prop derived from the feature flag and browser support status (`isFullSupport` from `useBrowserSupport` hook). The offset is only applied for fully supported browsers (Chrome, Brave, Edge) to ensure the banner displays correctly for limited support (Firefox) and unsupported browsers. Includes a TODO to remove the breakpoint logic when the flag is retired.
- **`apps/studio/src/App.tsx`**
  - — Added `<InfoBanner />` component to the app root, positioned after `ScrollToTop` and before the main `Background` component to ensure it appears at the top of the viewport.
- **`apps/studio/src/components/index.ts`**
  - — Added export for `InfoBanner` component to make it available for import throughout the app.
- **`apps/studio/src/pages/home/Home.tsx`**
  - — Made button `top` positioning breakpoint-aware (only adjusts on mobile where banner is in flow, not on desktop where banner is fixed). Removed unused `useTheme` import. Added `useBreakpoint` hook for responsive logic.
- **`apps/studio/src/common/hooks.ts`**
  - — Added `useMemo` to `useBreakpoint` hook return value to ensure referential stability and prevent unnecessary re-renders if the returned object is used in dependency arrays. Modified to ensure `useBreakpoint` hook is available for responsive breakpoint utilities used by `InfoBanner` and page components for conditional styling based on viewport size.
- **`packages/components/src/index.ts`**
  - — Ensured `Banner` component is properly exported from the components package (may have been reordered or formatting updated).
- **`apps/studio/src/pages/home/SideBar.tsx`**
  - — Added conditional `marginTop` (5 theme units) when banner is displayed on desktop to prevent content overlap with the fixed banner. Removed unused `useTheme` hook import to fix variable shadowing issue.
- **`apps/studio/src/pages/login/Login.tsx`**
  - — Added conditional `marginTop` (5 theme units) when banner is displayed on desktop to ensure proper spacing below the fixed banner.
- **`apps/studio/src/pages/signUp/SignUp.tsx`**
  - — Added conditional `marginTop` (2 theme units) when banner is displayed on desktop to maintain proper layout spacing.
- **`apps/studio/src/pages/signUp/Welcome.tsx`**
  - — Added conditional `marginTop` (3 theme units) when banner is displayed on desktop to maintain proper layout spacing.
- **`apps/studio/src/pages/signUp/Verification.tsx`**
  - — Added conditional `marginTop` (4 theme units) when banner is displayed on desktop for consistent spacing.
- **`apps/studio/src/pages/forgotPassword/ForgotPassword.tsx`**
  - — Added conditional `marginTop` (5 theme units) when banner is displayed on desktop to maintain proper page layout.
- **`packages/utils/src/lib/constants.ts`**
  - — Added browser support constants `FULL_SUPPORT_BROWSERS` and `LIMITED_SUPPORT_BROWSERS` to centralize browser support configuration. These constants define which browsers have full support (Chrome, Brave, Edge) and limited support (Firefox), making it easier to maintain and update browser support lists in a single location. The constants use `as const` for type safety and are consumed by the `useBrowserSupport` hook.
- **`packages/utils/src/lib/hooks.ts`**
  - — Added new `useBrowserSupport` hook that detects browser support level and returns `isFullSupport` and `isLimitedSupport` flags. The hook uses the centralized browser support constants (`FULL_SUPPORT_BROWSERS` and `LIMITED_SUPPORT_BROWSERS`) from `constants.ts` and `react-device-detect` to identify browser types. This hook centralizes browser detection logic and is used by `Background.tsx` and `UnsupportedBrowserBanner.tsx` to conditionally apply layout offsets and display browser support warnings, ensuring consistent browser detection across the codebase.
- **`packages/components/src/lib/UnsupportedBrowserBanner.tsx`**
  - — Refactored to use the centralized `useBrowserSupport` hook from `@newm-web/utils` instead of inline browser detection logic. Removed duplicate browser support arrays and `browserName` import from `react-device-detect`, replacing them with the shared hook. This improves code maintainability and ensures consistent browser detection logic across components. The component now uses `isFullSupport` and `isLimitedSupport` from the hook to determine banner visibility and messaging.

</details>

---

## Impact

- **User Experience**: Users will now see a prominent banner at the top of the page (on desktop) informing them about service discontinuation when the feature flag is enabled. The banner is non-intrusive and provides a clear call-to-action link.
- **Layout Adjustments**: The `Background` component applies a conditional top offset to its overlay when the banner flag is on (960px–xl), providing a single point of layout control for content under the fixed banner. Page-level conditional `marginTop` on Login, SignUp, SideBar, etc. remains for per-page spacing where needed. Together, these ensure content is not obscured.
- **Component API Enhancement**: The `Banner` component now supports more flexible styling through `sx` and `titleSx` props, making it more reusable and consistent with MUI patterns.
- **Code Reusability**: Extracted `OfficialStatementCTA` component to avoid duplication of the official statement message across multiple locations, improving maintainability. Centralized browser detection logic into `useBrowserSupport` hook and browser support arrays into constants (`FULL_SUPPORT_BROWSERS`, `LIMITED_SUPPORT_BROWSERS`), eliminating duplicate browser support arrays and detection logic across components. This provides a single source of truth for browser support configuration.
- **Testing Coverage**: Added comprehensive unit tests for the `Banner` component, improving code quality and maintainability.
- **Performance**: Minimal impact - banner only renders when feature flag is enabled, and uses efficient conditional rendering.
- **Accessibility**: Banner uses semantic HTML and proper link attributes (`rel="noopener noreferrer"`, `target="_blank"`) for external links.

---

## Testing

- **Unit Tests**: Added 22 test cases for the `Banner` component covering:
  - Basic rendering with title and description (string and ReactNode)
  - All prop variations (background, textColor, fixed, fullWidth, textAlign, sx, titleSx)
  - Default values and theme fallbacks
  - Conditional rendering logic
  - Style application and prop merging

- **Manual Testing**:
  - Verified banner appears/disappears based on `webStudioDisableDistributionAndSales` flag
  - Confirmed fixed positioning works correctly on desktop viewports
  - Tested responsive behavior on mobile (banner not fixed)
  - Verified all page layouts maintain proper spacing when banner is displayed
  - Confirmed external link opens in new tab with proper security attributes
  - Tested banner visibility across all modified pages (Login, SignUp, Home, etc.)

- **Test Commands**:

```bash
# Run unit tests for the Banner component
npx nx test components --testPathPattern="Banner"

# Run all tests for the components package
npx nx test components
```

---

## Related Issues

- Closes [STUD-575](https://projectnewm.atlassian.net/browse/STUD-575)

---

## Dependencies

No new dependencies.

---

## Additional Notes

- All spacing adjustments use MUI theme units for consistency and responsive behavior.
- The banner is conditionally rendered based on both the feature flag and viewport size (fixed only on desktop).
- The implementation follows existing patterns in the codebase for feature flag usage and component composition.
- The `Banner` component now includes `data-testid="banner"` to support testing without direct DOM node access, following Testing Library best practices.
- Extracted `OfficialStatementCTA` component to centralize the official statement message and link, making it easier to update the message or link URL across all usages.
  - **🚨 The link uses `NEWM_IO_URL` pointing to `https://newm.io/`. Once Marketing finalizes with the "Official statement" webpage, we will need to append the webpage path.**

**Design Decision - Banner Height/Padding:**

- The banner padding was intentionally kept smaller than the Figma design specifications to minimize layout disruption across pages. Increasing the padding to match Figma exactly would require additional conditional `marginTop` adjustments on numerous page elements (titles, headers, etc.), creating more maintenance overhead and potential inconsistencies. The current implementation provides sufficient visual prominence while maintaining a cleaner, more maintainable codebase. The banner's current appearance is considered sufficient for its purpose of communicating the service disruption message. Nevertheless, I'm open to discussing adjustments if design alignment is prioritized or if feedback suggests the banner needs more visual weight. **Background overlay offset**: `Background.tsx` applies a responsive top offset (`theme.spacing(2)`) to the main content overlay when the banner is shown (flag on and browser is fully supported, viewport 960px–xl), so all routes inside `Background` are shifted down in one place. Per-page `marginTop` values are retained where specific screens need additional spacing below the banner.

**Code Quality Note - Theme Access Standardization:**

- During implementation, a theme variable shadowing issue was identified and fixed in `SideBar.tsx` (removed unused `useTheme()` hook call that was shadowing the imported `theme` object). This highlights an existing inconsistency across the codebase where some components use the imported theme object directly (`import theme from "@newm-web/theme"`), while others use the MUI theme hook (`const theme = useTheme()`). This inconsistency can lead to confusion and potential bugs. **Future work**: We should consider standardizing theme access patterns across the codebase to improve maintainability and reduce the risk of using the wrong theme object.

--END--

[STUD-575]: https://projectnewm.atlassian.net/browse/STUD-575?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[STUD-575]: https://projectnewm.atlassian.net/browse/STUD-575?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/components/src/lib/test/Banner.test.tsx [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    expect(banner).toBeInTheDocument();
+    // MUI Box applies textAlign prop, verified by component rendering
+  });
+
+  it("defaults textAlign to center", () => {
+    renderWithContext(<Banner title="Test" />);
+
+    const banner = screen.getByTestId("banner");
+    expect(banner).toBeInTheDocument();
+  });
```
[the remark this answers is not in the document]
Comment (ITurres) [reply packages/components/src/lib/test/Banner.test.tsx:116]: ^ Outdated ^

File: apps/studio/src/common/hooks.ts [hunk 2]
```diff
@@ -35,3 +36,11 @@ export const useAuthenticatedRedirect = () => {
...
   }, [isLoggedIn]);
 };
+
+export const useBreakpoint = () => {
+  const theme = useTheme();
+  const isDesktop = useMediaQuery(theme.breakpoints.up("md"));
+  const isMobile = !isDesktop;
+
+  return { isDesktop, isMobile };
+};
```
[the remark this answers is not in the document]
Comment (ITurres) [reply apps/studio/src/common/hooks.ts:48]: ^ Outdated ^

File: packages/components/src/lib/Banner.tsx
Comment (dmkirshon) [inline packages/components/src/lib/Banner.tsx:1]: The banner also covers the `packages/elements/src/lib/Modal.tsx` closing icon, giant X in the top right. An example of this is if you set your wallet to `Mainnet` in garage/staging. I think most of the other examples like `viewPDF.tsx` for the Contract Agreement document review will be inaccessible with the flag.

With Banner:
Image


Before Banner:
Image

Comment (dmkirshon) [inline packages/components/src/lib/Banner.tsx:1]: There is also doesn't play well with the unsupported Browser Banner, `packages/components/src/lib/UnsupportedBrowserBanner.tsx`, the info banner is covered up and the top margins are offset. Once the unsupported Browser Banner is dismissed the background goes back to normal with the info banner. 

Image

File: apps/studio/src/components/OfficialStatementCTA.tsx [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import { NEWM_IO_URL } from "../common";
+
+interface OfficialStatementCTAProps {
+  readonly children?: ReactNode;
+  readonly href?: string;
+  readonly linkText?: string;
+}
+
+const OfficialStatementCTA: FunctionComponent<OfficialStatementCTAProps> = ({
+  href = `${NEWM_IO_URL}sunset/`,
```
Comment (dmkirshon) [inline apps/studio/src/components/OfficialStatementCTA.tsx]: Would rather see this in the constants.

**Review by dmkirshon — APPROVED** [review]
Looks good, please update the sunset NEWM url before merging in. The other two comments can be addressed in separate tickets.

File: apps/studio/src/components/OfficialStatementCTA.tsx (hunk 3 above)
Comment (ITurres) [reply apps/studio/src/components/OfficialStatementCTA.tsx]: Agree! Done ✅

File: packages/components/src/lib/Banner.tsx
Comment (ITurres) [reply packages/components/src/lib/Banner.tsx:1]: Thanks @dmkirshon! Done🚀 ✅

Comment (ITurres) [reply packages/components/src/lib/Banner.tsx:1]: Thanks @dmkirshon! Done🚀 ✅

File: packages/utils/src/lib/hooks.ts [hunk 4]
```diff
@@ -341,3 +343,16 @@ export const useBetterMediaQuery = (mediaQueryString: string) => {
...
+  const limitedSupportBrowsers = ["Firefox"];
+
+  const isFullSupport = fullSupportBrowsers.includes(browserName);
+  const isLimitedSupport = limitedSupportBrowsers.includes(browserName);
+
+  return {
+    isFullSupport,
+    isLimitedSupport,
+  };
+};
```
Comment (dmkirshon) [inline packages/utils/src/lib/hooks.ts:363]: Is the plan to update `packages/components/src/lib/UnsupportedBrowserBanner.tsx` at a later time to maintain consistency between the repeated support code?

Comment (ITurres) [reply packages/utils/src/lib/hooks.ts:363]: Just done it 😃. Thanks David!
